### PR TITLE
Re-write the grpc service APIs

### PR DESCRIPTION
**Pull request by ethanrublee**
Sketching a simplification of the grpc api

### CODE REVIEW

File: protos/farm_ng/api/service.proto [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+message GetServiceStateRequest {
+}
+
+message GetServiceStateReply {
+  ServiceState state = 1;
+  int pid = 2;
+  double uptime = 3;
+}
+
+service Service {
```
Comment (ethanrublee) [inline protos/farm_ng/api/service.proto:41]: Note multiple services can be implemented on the same port with grpc, so we can compose this generic interface with each of our services easily.  Having one Service client stub, will make the launcher not need to know about the specifics of a service.

File: protos/farm_ng/api/service.proto [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  rpc getServiceState(GetServiceStateRequest) returns (GetServiceStateReply) {}
```
Comment (edgarriba) [inline protos/farm_ng/api/service.proto:42]: Don't we need a restartService? For example the camera sometimes we need to manually restart via the GUI. If we target e.g. a ServicesManager we should have a minimum interaction to secure a good state of the services.

Comment (ethanrublee) [reply protos/farm_ng/api/service.proto:42]: Looking at this, we don't need to restart the service through the grpc, that's the responsibility of the launcher, which can start and start the process.  Also, it seems not very safe if any client can restart the service.

Comment (edgarriba) [reply protos/farm_ng/api/service.proto:42]: ServicesManager would be a master service that decouples from the launcher gui, something we discussed several times to do with @christopherbradski that also gives us easy discoverability of the state of the services with minimal interaction.

Comment (ethanrublee) [reply protos/farm_ng/api/service.proto:42]: Yes that might be a good idea to add a services management API, but currently, this is focused on improving our existing services' APIs.

File: protos/farm_ng/canbus/canbus.proto [hunk 3]
```diff
@@ -74,39 +64,14 @@ message RawCanbusMessages {
...
-  string state_name = 1;
-  CanbusServiceState state = 2;
-  ReplyStatus status = 3;
+message GetIfaceStateReply {
+  string can_iface = 4;
+  int32 bit_rate = 5;
+  CanbusIfaceState iface_state = 6;
+  int32 tec = 7;
+  int32 rec = 8;
+  float bus_load = 9;
```
Comment (Hackerman342) [inline protos/farm_ng/canbus/canbus.proto:76]: Captured a follow up in workspace/ to actually use these: https://github.com/farm-ng/workspace/issues/973

File: protos/farm_ng/state_estimator/state_estimator.proto [hunk 4]
```diff
@@ -12,47 +12,43 @@
...
 
-message RobotDynamicsResult {
+message RobotDynamicsReply {
     // Timestamp from farm-ng-core
     // SE3d types from Sophus
-    Timestamp stamp = 1;  // this contains which clock
+    farm_ng.core.proto.Timestamp stamp = 1;  // this contains which clock
     sophus.proto.Se3F64 odom_pose_robot = 2;
     // Rate of the robot in the robot's frame
     RobotRate robot_rate = 3;
```
Comment (Hackerman342) [inline protos/farm_ng/state_estimator/state_estimator.proto:54]:
```suggestion
    farm_ng.core.proto.Timestamp stamp = 1;  // this contains which clock
    sophus.proto.Se3F64 odom_pose_robot = 2;
    // Rate of the robot in the robot's frame
    RobotRate robot_rate = 3;
    AmigaControlState amiga_state = 4;
```

File: protos/farm_ng/state_estimator/state_estimator.proto [hunk 5]
```diff
@@ -12,47 +12,43 @@
...
+// message ControlRobotDynamicsReply {
 //     // Pose from state estimator to controller
 // }
 
 message RobotDynamicsRequest {
     string device = 1;
 }
 
-message RobotDynamicsResult {
+message RobotDynamicsReply {
```
Comment (Hackerman342) [inline protos/farm_ng/state_estimator/state_estimator.proto:48]: While we're here, nice to address #30  
```suggestion
enum AmigaControlState {
    STATE_BOOT = 0;
    STATE_MANUAL_READY = 1;
    STATE_MANUAL_ACTIVE = 2;
    STATE_CC_ACTIVE = 3;
    STATE_AUTO_READY = 4;
    STATE_AUTO_ACTIVE = 5;
    STATE_ESTOPPED = 6;
}

message RobotDynamicsReply {
```

**Review by Hackerman342 — APPROVED** [review]
Will require a lot on the `workspace/` side to integrate. Happy to take that on if you haven't already started.

File: protos/farm_ng/api/service.proto [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  // Idle, waiting for clients
+  IDLE = 3;
+  // Service isn't running at all
+  UNAVAILABLE = 4;
+  // Service is in some error state
+  ERROR = 5;
+}
+
+message GetServiceStateRequest {
+}
```
[the remark this answers is not in the document]
Comment (ethanrublee) [reply protos/farm_ng/api/service.proto:33]: The idea here was just to have something grounded in the which process is publishing the status; which could be useful, especially when understanding if a service is restarting frequently.

File: protos/farm_ng/state_estimator/state_estimator.proto (hunk 4 above)
Comment (ethanrublee) [reply protos/farm_ng/state_estimator/state_estimator.proto:54]: It doesn't feel like the AmigaControlState belongs in the state estimation service which could be hardware independent.  I would suggest that we add AmigaControlState and AmigaTPDO1, AmigaRPDO1, as a separate service interface (possibly implemented in the same process as the canbus service).  This way a client that doesn't care about the specifics of the Amiga can just use the state estimation API.

File: protos/farm_ng/state_estimator/state_estimator.proto (hunk 5 above)
Comment (ethanrublee) [reply protos/farm_ng/state_estimator/state_estimator.proto:48]: Ideally for enums we have an UNKNOWN=0, or UNSPECIFIED=0 - https://developers.google.com/protocol-buffers/docs/style#enums

File: protos/farm_ng/api/service.proto [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package farm_ng.api.proto;
+
+enum ServiceState {
+  UNKNOWN = 0;
+  STOPPED = 1;
```
Comment (edgarriba) [inline protos/farm_ng/api/service.proto:21]: not very clear what does it mean that a service is stopped vs unavailable from client perspective

File: protos/farm_ng/canbus/canbus.proto [hunk 8]
```diff
@@ -19,45 +19,35 @@ syntax = "proto3";
...
-  rpc streamCanbusMessages(StreamCanbusRequest)
-      returns (stream StreamCanbusReply) {}
-  rpc sendCanbusMessage(stream SendCanbusMessageRequest)
-      returns (SendCanbusMessageReply) {}
-  rpc getServiceState(GetServiceStateRequest) returns (GetServiceStateResult) {}
-  rpc startService(StartServiceRequest) returns (StartServiceResult) {}
-  rpc stopService(StopServiceRequest) returns (StopServiceResult) {}
-  rpc pauseService(PauseServiceRequest) returns (PauseServiceResult) {}
+  rpc readCanbusMessages(ReadCanbusMessagesRequest)
+      returns (stream ReadCanbusMessagesReply) {}
```
Comment (edgarriba) [inline protos/farm_ng/canbus/canbus.proto:23]: one thought discussed with Kyle is that the need both `readCanbusMessages` and `sendCanbusMessage` ? 
we could simplify to the single call with bidirectional streaming

File: protos/farm_ng/controller/controller.proto [hunk 9]
```diff
@@ -22,19 +22,8 @@ import "sophus/lie.proto";
 package farm_ng.controller.proto;
 
 service ControllerService {
-  // called from the gui
-  // Reports progress back to gui until it achieves the goal
-  // TODO how to cancel the request (possibly by cancelling the response stream? or using the input as a stream)
   rpc moveToGoalPose(MoveToGoalPoseRequest)
```
Comment (edgarriba) [inline protos/farm_ng/controller/controller.proto:25]: same as in canbus, do we want to request as stream ?

File: setup.cfg [hunk 10]
```diff
@@ -41,6 +41,7 @@ package_dir =
     = py
 packages =
     farm_ng
+    farm_ng.api
```
Comment (edgarriba) [inline setup.cfg:44]:
```suggestion
    farm_ng.service
```
api to generic to me indicating the whole thing

**Pull request closed without merging**